relay-general/src/user_agent.rs

```diff
@@ -83,51 +84,119 @@ pub fn parse_os(user_agent: &str) -> OS {
 ///
 /// Useful for the scenarios where you will use either information from client hints if it exists,
 /// and if not fall back to user agent string.
-#[derive(Default, Debug)]
-pub struct RawUserAgentInfo<'a> {
+#[derive(Default, Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct RawUserAgentInfo<S: Default + AsRef<str>> {
     /// The "old style" of a single UA string.
-    pub user_agent: Option<&'a str>,
+    pub user_agent: Option<S>,
     /// User-Agent client hints.
-    pub client_hints: ClientHints<'a>,
+    pub client_hints: ClientHints<S>,
 }
 
-/// The client hint variable names mirror the name of the "SEC-CH" headers, see
-/// '<https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers#user_agent_client_hints>'
-#[derive(Default, Debug)]
-pub struct ClientHints<'a> {
-    /// The client's OS, e.g. macos, android...
-    pub sec_ch_ua_platform: Option<&'a str>,
-    /// The version number of the client's OS.
-    pub sec_ch_ua_platform_version: Option<&'a str>,
-    /// Name of the client's web browser and its version.
-    pub sec_ch_ua: Option<&'a str>,
-    /// Device model, e.g. samsung galaxy 3.
-    pub sec_ch_ua_model: Option<&'a str>,
+impl<S: AsRef<str> + Default> RawUserAgentInfo<S> {
+    /// Checks if key matches a user agent header, in which case it sets the value accordingly.
+    /// Extracted out from from_headers so that it can be called from another crate with different
+    /// header-types. Perhaps we could make it generic over different header types instead.
```

Comment:
```suggestion
    /// Checks if key matches a user agent header, in which case it sets the value accordingly.
    ///
    /// TODO(tor): make it generic over different header types.
```
> Extracted out from from_headers so that it can be called from another crate with different header-types.

This sentence suits as a comment in the PR, but I don't think it's useful in the docstrings.

> Perhaps we could make it generic over different header types instead.

This suggestion is probably personal, but I'd leave this as a TODO.